오픈마켓 [Step 2] Gundy, Wonbi

wonbi92 opened the pull request:
@wonhee009 
안녕하세요 라자냐! 
이번 스텝은 처음 써보는 네트워크 작업에 더불어 처음 써보는 컬렉션뷰라서 생각보다 많이 헤맨 것 같습니다! 이상한 부분이 있다면 다 알려주세요!

## ✓ 진행단계 : Step-2

## 📝 STEP 진행 중 경험하고 배운 것
- 컬렉션 뷰 사용하기
    ☑️ 커스터마이징 셀 구현하기
    ☑️ 리스트와 그리드 모양의 컬렉션 뷰 구현하기
- 이미지 비동기로 처리하기
    ☑️ 이미지를 서버에서 파싱하는 과정을 비동기로 처리하기
    ☑️ race condition 해결하기
- UISegmentedControl 사용하기
- 이미지 캐싱하기
    ☑️ 서버를 통해서 전달받은 데이터를 로컬에 캐시하기

## ⚒️ 코드 구현내용 
<details>
<summary>구현내용</summary>
    
###  구현
- Extension
    - DecodingError
        - `errorDescription`을 사용해 상황에 맞는 에러 메세지를 출력하도록 하였습니다.
    - Collection
        - `subscript`를 이용해 첨자 문법으로 값에 접근 시 런타임에러가 나지 않도록 하였습니다.
    - Int
        - 값이 0인지 확인하는 `isZero`, 숫자의 자리수를 체크하는 `decimal` 프로퍼티를 가지도록 하였습니다.
- OpenMarket
    - Product
        - 이제 `Product`는 `Hashable`을 채택합니다.
    - ImageCacheManager
        - 이미지를 캐싱하기 위한 싱글톤 객체입니다.
- Controller
    - ProductsViewController
        - 앱 실행시 나오는 첫 화면을 컨트롤 합니다.
        - 데이터를 파싱하고 이를 각 컬렉션 뷰에 전달합니다.
        - segmentedControl의 값이 바뀔 때 마다 각각의 컬렉션 뷰를 보여주도록 화면을 전환합니다.
    - AddProductViewController
        - 다음 스텝에서 추가될 새로운 상품을 등록하는 화면을 컨트롤합니다.
- DummyData
    - ListCollectionViewCell
        - 리스트 형태의 컬렉션 뷰에서 사용하는 셀입니다.
        - 리스트 형태로 커스터마이징 된 셀을 그립니다.
    - GridCollectionViewCell
        - 그리드 형태의 컬렉션 뷰에서 사용하는 셀입니다.
        - 그리드 형태로 커스터마이징 된 셀을 그립니다.
    
</details>


## 💭 고민한 부분

### 코드로만 화면 구현하기
- 이번 프로젝트는 유독 스토리보드로 구현하기가 더 쉬웠을 것 같습니다. 하지만 양쪽 모두 할 줄 알아야하기 때문에 이번 프로젝트에서는 코드로만 화면을 구현하는 것을 목표로 했습니다.
- 이 과정에서 초기 설정으로 필요한 부분들을 AppDelegate 및 SceneDelegate에 작성하여 진행했습니다.

### 안전한 Collection 사용을 위한 extension
- API를 통해 데이터를 요청할 때 한 번에 20개씩 요청을 하고, 마지막에 가서는 20개보다 적게 응답이 오기 때문에 잘못된 인덱스에 접근할 가능성이 있습니다.
- 이를 subscript 문법에 대한 메서드 추가로 안전하게 값에 접근하도록 하였습니다.

### 투명한 네비게이션 바 문제 해결하기
- 앱 구동시 네비게이션 바가 투명해져서 네비게이션 바 아이템들이 둥둥 떠다니는 모습이 되어버렸습니다. 이는 iOS가 15버전으로 업데이트 된 후 네비게이션 바가 확장되면서 생긴 문제였습니다. 시뮬레이터의 iOS버전은 15 였고, 시뮬레이터에서 네비게이션 바가 투명해 진 것입니다. 
- 저희가 원했던 방향성은 네비게이션 바가 항상 불투명하게 자리를 잡는 것이기 때문에 `UINavigationBarAppearance`의 `configureWithDefaultBackground()` 인스턴스 메서드를 사용해서 항상 불투명하게 나오도록 처리하였습니다.

### 이미지를 가져올 때 비동기로 인한 race-condition발생 문제 해결
- 이미지를 서버에서 가져오는 로직의 경우, 이미지의 크기가 커 불러오는 작업을 메인 스레드에서 진행하면 스크롤을 내릴 때 마다 화면이 버벅이게 되어 쾌적한 사용자 경험을 제공하지 못하게 됩니다. 
- 따라서 이 로직을 메인 스레드가 아닌 다른 스레드에서 진행하도록 하여 메인 스레드는 받아온 이미지를 띄우기만 하도록 로직을 짜보았습니다.
- 문제는 이 작업을 비동기적으로 진행하다보니, 이미지를 불러오기 위한 `thumbnail`에 접근하는 과정에서 race-condition이 발생하는 것입니다. 때문에 스크롤을 빠르게 내릴 수록 이미지가 계속 바뀌는 상태가 되었습니다.
- 이를 해결하기 위해 이미지를 불러오기 전에 셀이 가지고 있는 `product`와 디스패치 큐로 작업을 넘길 때 캡쳐한 `product`가 일치하는지 확인하는 로직을 추가하였고, `product`가 비교 가능한 상태가 될 수 있도록 `Hashable`을 채택하도록 처리하였습니다.

### 이미지 로딩중임을 사용자에게 알리기
- 이미지를 서버에서 가져오는 로직은 텍스트를 가져오는 것과는 다르게 시간이 걸리는 작업입니다. 따라서, 이미지의 파싱이 끝날 때 까지 사용자는 이미지가 없는 셀을 보다가 이미지가 나중에 나타나는 UI를 보게 될 것입니다. 이는 좋은 사용자 경험이 아니라 판단하여 이미지의 파싱이 끝날 때 까지 로딩중임을 알리는 시각적 정보가 필요하였습니다.
- 그래서 이미지와 정확히 같은 위치에 `UIActivityIndicatorView`를 추가하여 이미지가 파싱되어 이미지뷰에 할당될 때까지 로딩중임을 알렸습니다.
- 이미지가 할당된 후에는 `UIActivityIndicatorView`의 애니메이션을 멈추고 보이지 않게 바꾸도록 하였습니다.

### 이미지 캐싱하기
- 매번 셀을 dequeue할 때마다 이미지를 서버에서 불러오는 것은 자원 낭비가 너무 심하다 판단되었습니다. 한번 불러온 이미지를 캐싱하여 이 문제를 해결하고자 했습니다.
- NSCache<NSString, UIImage>타입의 싱글턴 객체를 갖는 ImageCacheManager 클래스를 구현하였습니다.
- 이를 통해 한 번 사용된 이미지는 캐싱하여 성능을 향상시킬 수 있도록 구현하였습니다.

## ❓ 조언을 얻고 싶은 부분

### segmentedControl과 화면 전환
- 현재 저희는 두 개의 컬렉션 뷰를 가지고 서로 isHidden 상태를 전환하는 방식으로 화면을 변경하고 있습니다. 하나의 컬렉션 뷰를 가지고 레이아웃만 변경해서 화면을 바꾸는 것이 더 적절한 방법인지에 대한 라자냐의 생각이 궁금합니다. 또한 현재는 segmentedControl의 item이 2개지만 더 많은 경우에는 또 어떤 방법을 취하는 것이 좋을까요?

### 스크롤 동기화
- 저희는 segmentedControl을 통한 화면 전환에서 두 화면이 서로 같은 상품을 보여주도록 스크롤 동기화가 있으면 좋겠다고 생각했습니다.
- 특정 인덱스패스로 스크롤 해주는 메서드를 사용하고자 cellForItemAt이 호출될 때, indexPath 전달인자를 뷰컨트롤러의 프로퍼티에 저장했다가 화면전환시 해당 인덱스패스로 스크롤하는 `scrollToItem(at:at:animated:)` 메서드를 호출해봤습니다. 하지만 이 방법은 가장 마지막에 반환된 cell에 대한 인덱스로 스크롤되기 때문에 스크롤 방향에 따라서 `at: UICollectionView.ScrollPosition`을 적절히 바꿔줘야 했고, 완벽하게 작동하지도 않았습니다.
- 두 컬렉션뷰 레이아웃의 아이템 사이즈의 비율과 근사하게 contentOffset에 2.1을 곱하거나 나눠서 위치를 잡아주는 방법도 시도해봤습니다. 얼추 비슷하게 스크롤되긴 했습니다.
- 하지만 더 적절하게 스크롤을 동기화 시키려면 화면에 보이는 첫 번째 셀의 인덱스 패스를 가지고 `scrollToItem(at:at:animated:)`에 `UICollectionView.ScrollPosition.top`와 함께 전달하는 것이 좋겠다고 생각했습니다. 헌데 저희는 화면에 보이는 첫 번째 셀의 인덱스 패스를 어떤 방법으로 알아낼 수 있는지 모르겠습니다.
- 현재는 두 컬렉션 뷰의 스크롤을 동기화하지 않았습니다.
- 첫 번째 셀의 인덱스 패스를 알아내는 방법이나, 더 좋은 스크롤 동기화에 대한 아이디어가 있다면 조언을 부탁드립니다!

### 모던 컬렉션 뷰사용
- 컬렉션 뷰를 구현하는 방법을 고민하면서, 처음에는 모던 컬렉션 뷰를 사용하여 구현하였습니다. 하지만, 모던 컬렉션 뷰가 어떻게 동작하는지 정확하게 이해하진 못했고, 그저 예제를 보면서 카피코딩만 하고 있는거 같아 모던 컬렉션 뷰가 아닌 기존 컬렉션 뷰로 구현하게 되었습니다. 이 부분은 좀 더 공부한 후에 적용시키자고 생각하였습니다.
- 저희가 궁금한 것은 이 모던 컬렉션 뷰와 기존 컬렉션 뷰 둘 중에 어떤 방식이 현업에서 더 많이 사용되는 지 궁금합니다. 회사마다 다르겠지만 라자냐의 의견이 궁금합니다!

### available
- 모던 컬렉션뷰를 사용해보았을 때 iOS 14버전부터 사용할 수 있다는 available이 뜨게 되었습니다.
- 이런 경우 14버전보다 낮은 버전에서는 해당 기능을 사용할 수 없기 때문에 버전 분기처리를 통한 같은 기능 구현을 필요로 할텐데요! 저희가 궁금한 점은 버전 분기처리는 보통 몇 버전정도 이전까지 구현하는 것이 보편적인가 하는 것이었습니다.
- 라자냐는 현재 몇 버전까지 구현해주는 것이 좋다고 생각하시나요?

wonbi92 committed:
feat : 코드로 뷰를 그리기 위한 세팅

- SceneDelegate의 scene메서드에 메인화면 뷰 컨트롤러를 루트 뷰 컨트롤러로 선언하는 로직 추가

Gundy93 committed:
feat: navigationController 추가 및 navigationItem 추가

- SceneDelegate에서 window?.rootViewController을 navigationController로 변경
- segmentedControl 추가하여 navigationItem.titleView에 할당
- addProductButton 추가하여 navigationItem.rightBarButtonItem 할당

wonbi92 committed:
feat : collectionView 구현

- list 모양의 collectionView 구현
- 레이아웃 타입을 결정해주는 LayoutType 열거형 구현
- list 모양에 맞는 ListCollectionViewCell 커스텀 셀 구현

Gundy93 committed:
feat : 유저의 스크롤에 맞춰 데이터를 추가로 불러오도록 로직 구현

- Collection을 extension하여 subscript 문법을 통해 인덱스에 접근할 때 옵셔널 값을 뱉게끔 구현
- JSON 형식에 맞게 모델 수정
- ProductsViewController의 프로퍼티 productList를 productLists로 수정
- ProductsViewController extension을 통해 UICollectionViewDelegate 채택 및 scrollViewDidScroll 내부 구현을 통해 데이터 추가 로딩 구현

wonbi92 committed:
feat : List와 Grid 방식 컬랙션 뷰 구현

- GridCollectionView에서 사용하는 커스텀 GridCollectionViewCell 구현
- gridCollectionView프로퍼티 생성 및 오토레이아웃 지정
- segmentedControl의 선택된 항목에따라 List와 Grid가 보여지도록 설정하는 changeLayout 메서드 구현
- segmentedControl에 addTarget해주는 메서드 구현

Gundy93 committed:
refactor : List와 Grid의 스크롤 동기화

wonbi92 committed:
feat : ListCollectionViewCell의 UI 구현

- prepareForReuse메서드 구현
- 뷰를 그리고 데이터를 주입하는 setupCellConstraints및 configureCell메서드 구현
- 데이터를 주입하는데 필요한 setupPrice, setupStock 메서드 구현
- FormatConverter의 메서드를 static으로 변경 및 numberFormatter 구현
- Int타입에 0인지 확인하는 isZero 및 자리수를 확인하는 decimal 프로퍼티 추가

Gundy93 committed:
feat: gridCollectionView용 커스텀 셀 구현 및 이미지 로딩 로직 구현

- 세부 inset 조정

wonbi92 committed:
refactor : GridCollectionView 오토레이아웃 수정 및 이미지 네트워킹 로직 수정

- GridCollectionView에서 사진의 오토레이아웃 오류 수정
- 이미지 네트워킹 과정에서 생기는 레이스 컨디션 문제 해결

Gundy93 committed:
feat: Cell의 Layer에 border 추가

- scroll 동기화 로직 정리

wonbi92 committed:
feat : + 버튼을 누르면 빈 창이 나타나도록 구현

- addProductButton의 액션과 타겟 지정
- addNewProduct 메서드 구현
- 상품명이 없어도 productName 레이블이 생성되도록 로직 수정
- navigationBarAppearance를 이용한 불투명한 네비게이션바가 나오도록 수정
- AddProductViewController 생성

Gundy93 committed:
feat: 이미지 캐싱 구현

- ImageCacheManager 클래스의 싱글턴객체를 통한 이미지 캐싱 구현

wonbi92 committed:
chore : 주석 및 개행, 인덴트 정리

wonhee009 commented on OpenMarket/OpenMarket/ProductsViewController.swift:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+    
+    private var networkManager = NetworkManager()
```
이후에 바뀔 여지가 있는 프로퍼티인가요?

wonbi92 replied:
내부 세션이 변경될 수 있어서 var로 선언했었는데 생각해보니 클래스이기 때문에 let이 맞는거 같습니다! let으로 수정하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/ProductsViewController.swift:
```diff
@@ -0,0 +1,221 @@
+    }()
+    
+    private let segmentedControl: UISegmentedControl = {
+        let segmentedControl = UISegmentedControl(items: ["LIST", "GRID"])
```
"GRID, LIST"로 순서가 변경된다면 `LayoutType`의 index도 적절히 변경해줘야 이슈가 없을거 같은데요.
순서/index에 대한 부분을 한 군데에서 관리할 수 있도록 하는건 어떨까요?
```swift
enum LayoutType: Int, CaseIterable {
   case list = 0
   case grid
  
   var text: String {
      switch self {
         case .list:
            return "LIST"
         ...
      }
   }
}
```

이렇게 선언한다면 `sorted`이용하면 순서도 보장할 수 있을거 같고, 수정이 있을 때에는 enum만 수정해주면 될거 같습니다.

wonbi92 replied:
이런 방법은 생각하지 못했었네요! 수정해보았습니다!

wonhee009 commented on OpenMarket/OpenMarket/ProductsViewController.swift:
```diff
@@ -0,0 +1,221 @@
+    
+    private func addTarget() {
+        segmentedControl.addTarget(self, action: #selector(changeLayout(_:)), for: .valueChanged)
+        addProductButton.target = self
+        addProductButton.action = #selector(addNewProduct)
```
`segmentedControl`처럼 `.addTarget()`을 사용하지 않은 이유가 있나요?.?

wonbi92 replied:
```Swift
private let addProductButton: UIBarButtonItem = {
    let barButton = UIBarButtonItem(title: "+", style: .plain, target: self, action:  #Selector(addNewProduct))
    let attributes: [NSAttributedString.Key : Any] = [.font: UIFont.boldSystemFont(ofSize: 16)]
    barButton.setTitleTextAttributes(attributes, for: .normal)
    return barButton
}()
```
UIBarButtonItem에는 addTarget 메서드가 없고, 초기화시에 타겟을 지정할 수 있는데, 이렇게 초기화 과정에서 타겟을 지정하면 런타임 에러가 발생하는 문제가 있었습니다. 생성 시점이 문제일까 해서 lazy var로 선언해 보았지만 결과는 동일하였습니다. 결국 타겟과 액션을 직접 지정하는 방식으로 사용하게 되었습니다. 혹시 저희가 놓친 부분이 있을까요?

wonhee009 replied:
아핫 네넵 이해했습니다~

wonhee009 commented on OpenMarket/OpenMarket/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,196 @@
+        return stackView
+    }()
+    
+    override var reuseIdentifier: String? {
+        return "ListCell"
+    }
```
cell reuse할때 사용하려고 정의한거 같은데요.
사용하고 있나요? VC에서 "ListCell" 텍스트 입력해서 사용하고 있는거 같아 질문 드립니다.

wonbi92 replied:
UICollectionView를 초기화 하는 과정에서 register메서드를 통해 지정해주고, reuseIdentifier도 지정해주어야 정상 작동 하는줄 알았습니다. register메서드에서 지정되는거군요! 이제 필요없는 코드가 되어서 삭제하였습니다

wonhee009 commented on OpenMarket/OpenMarket/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,196 @@
+    }
+    
+    func configureCell(from product: Product) {
+        self.product = product
```
cell이 product를 프로퍼티로서 가지고 있어야하나요?

Gundy93 replied:
product를 전부 소유할 필요가 없어서 uuid를 통한 identifier를 만들어서 그 identifier간의 비교를 통한 로직으로 변경하였습니다.
```Swift
struct Product: Codable, Hashable {
let identifier: String = UUID().uuidString
}
```

wonhee009 commented on OpenMarket/OpenMarket/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,196 @@
+    
+    func configureCell(from product: Product) {
+        self.product = product
+        loadingView.startAnimating()
```
prepareForReuse()에서도 로딩 애니메이션 시작하지 않나요?

Gundy93 replied:
두 번 사용되고 있는 것을 인지하지 못했네요! 둘 다 시작 전을 정비하는 메서드다보니 그렇게 적은 것 같습니다. startAnimating 시점은 셀을 실제로 사용하는 시점인 configureCell이 더 적절한 것 같아 prepareForReuse에서 startAnimating을 삭제하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/ListCollectionViewCell.swift:
```diff
@@ -0,0 +1,196 @@
+            DispatchQueue.main.async { [weak self] in
+                if product == self?.product {
+                    self?.productImage.image = image
+                    self?.loadingView.stopAnimating()
```
cachedImage를 가지고 오면 로딩뷰 애니메이션을 중지하는거 같은데요. 중복 코드 아닐지 확인 부탁드립니다.

Gundy93 replied:
```Swift
if let cachedImage = ImageCacheManager.shared.object(forKey: cacheKey) {
    productImage.image = cachedImage
    loadingView.stopAnimating()
    loadingView.isHidden = true
    productImage.isHidden = false
    return
}
```
위 코드 진행에서 cachedImage를 가지고 오면 아래의 코드가 진행되지 않으므로 두 번 중지되는 것은 아니라고 생각합니다!
```Swift
DispatchQueue.main.async { [weak self] in
    guard product.identifier == self?.identifier else { return }
    
    self?.productImage.image = image
    self?.loadingView.stopAnimating()
    self?.loadingView.isHidden = true
    self?.productImage.isHidden = false
}
```

wonhee009 commented on OpenMarket/OpenMarket/FormatConverter.swift:
```diff
@@ -8,17 +8,30 @@
+        return formatter
+    }
+    
+    static func number(from double: Double) -> String {
+        if let result = numberFormatter.string(for: double) {
+            return result
+        }
+        return String(double)
+    }
```
number라는 네이밍이 너무 제네럴한거 같습니다.
`numberFormatter`의 numberStyle이 `.decimal`인데 이것만 숫자인건 아니니까요.
좀 더 명확한 네이밍이 되면 좋겠습니다.
(위의 date도 마찬가지)

Gundy93 replied:
`convertToDate(from:)`과 `convertToDecimal(from:)`으로 네이밍 변경하였습니다! date의 경우 저희가 사용하려는 목적이 Date 타입의 값이므로 convertToDate로 사용하고, 숫자의 경우 decimal style의 값이 필요했기 때문에 convertToDecimal로 설정하였습니다.

wonhee009 reviewed (changes requested):
수고하셨습니다. 코멘트 남겨두었으니 확인 부탁드려요:)

- 현재 스펙이라면 list, grid 모두 동일한 데이터를 사용하기도 해서 저는 하나의 collectionView로 처리했을거 같습니다. segment 이동시 화면이 움직이는 모션이 있다면 두개로 처리했을거 같습니다. 주어진 스펙따라 다를거 같네요.
- [UICollectionView 공식문서](https://developer.apple.com/documentation/uikit/uicollectionview)는 확인해보셨을까요? 여기에 visibleCell에 관한 프로퍼티/메서드가 있는거로 알고 있습니다.
- modern collectionView 최소 타겟이 14.0으로 알고 있는데요. 현업에서 어떤거를 더 많이 사용하냐는 좀 다른 얘기일거 같습니다. 서비스별로 최소 타겟이 달라서 그 부분을 충족하면 modern collectionVIew를 사용할 순 있겠죠. 하지만 기존에 collectionView였던거를 누군가가 modern collectionView를 바꿔줘야만 하기 때문에 collectionView의 비율이 높긴할거 같구요. 사실 modern collectionView가 완전 새로운 부분은 아녀서 collectionView만 잘 알아도 된다고 생각합니다. 필요하면 그때 필요한 개념을 보충하는거라고 생각하고요.
- 버전 분기는 팀별로 다르기 때문에 제가 대답해줄 수 있는 부분은 아닌거 같고요. 몇 버전까지 구현도 서비스별로 상황이 달라서 얘기해주기 어려운 부분인거 같습니다.

Gundy93 committed:
docs : README.md 업데이트

Gundy93 committed:
docs: README.md 수정

wonbi92 committed:
refactor : ProductsViewController와 CollectionView의 로직 수정

- 각 Cell에서 reuseIdentifier 삭제
- 각 Cell의 setupImage메서드에서 race condition을 체크하는 로직 수정
- LayoutType의 rawValue가 Int이고 CaseIterable을 채택하도록 수정

Gundy93 committed:
refactor : 제네럴한 네이밍 수정, Cell 내부 로직 변경

- Int+의 decimal 프로퍼티 삭제
- FormatConverter의 date, number를 각각 convertToDate, convertToDecimal으로 변경
- 각 셀에 identifier 프로퍼티 추가하고 기존 product 프로퍼티 삭제
- Product에 identifier 프로퍼티 추가

wonbi92 committed:
refactor : collectionView 하나로 통합

- 네이밍 수정
- ListCollectionViewCell의 오토레이아웃 수정
- 그리드와 리스트 각각의 FlowLayout 선언
- gridCollectionView 프로퍼티, makeLayout메서드 삭제

Gundy93 committed:
feat : 레이아웃 전환시 indexPathsForVisibleItems를 통한 스크롤 동기화 구현

wonbi92 committed:
refactor : borderLine 재설정

- collectionView의 Layout이 listLayout일 때 borderLine 설정
- 첫 화면에서 그리드로 넘어갈 시 3번째 아이템부터 보이던 문제 수정

wonhee009 reviewed (approved):
수고하셨습니다:)
